File: c-algorithm/src/compare-string.c
// 识别到的包含指令
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include "compare-string.h"

// 识别到的函数定义
// 函数: string_equal (line 29)
int string_equal(void *string1, void *string2)
{
	return strcmp((char *) string1, (char *) string2) == 0;
}

// 函数: string_compare (line 34)
int string_compare(void *string1, void *string2)
{
	int result;

	result = strcmp((char *) string1, (char *) string2);

	if (result < 0) {
		return -1;
	} else if (result > 0) {
		return 1;
	} else {
		return 0;
	}
}

// 函数: string_nocase_equal (line 51)
int string_nocase_equal(void *string1, void *string2)
{
	return string_nocase_compare((char *) string1, (char *) string2) == 0;
}

// 函数: string_nocase_compare (line 60)
int string_nocase_compare(void *string1, void *string2)
{
	char *p1;
	char *p2;
	int c1, c2;

	/* Iterate over each character in the strings */

	p1 = (char *) string1;
	p2 = (char *) string2;

	for (;;) {

		c1 = tolower(*p1);
		c2 = tolower(*p2);

		if (c1 != c2) {

			/* Strings are different */

			if (c1 < c2) {
				return -1;
			} else {
				return 1;
			}
		}

		/* End of string */

		if (c1 == '\0')
			break;

		/* Advance to the next character */

		++p1;
		++p2;
	}

	/* Reached the end of string and no difference found */

	return 0;
}

File: comparisons/MinsC2Rust/c_algorithm/Output/test_project/src/compare_string.rs

pub fn string_nocase_compare(string1: &str, string2: &str) -> i32 {
    let mut p1 = string1.chars();
    let mut p2 = string2.chars();

    loop {
        let c1 = p1.next().map(|c| c.to_ascii_lowercase());
        let c2 = p2.next().map(|c| c.to_ascii_lowercase());

        match (c1, c2) {
            (Some(c1), Some(c2)) => {
                if c1 != c2 {
                    return if c1 < c2 { -1 } else { 1 };
                }
            }
            (Some(_), None) => return 1,
            (None, Some(_)) => return -1,
            (None, None) => return 0,
        }
    }
}

pub fn string_compare(string1: &str, string2: &str) -> i32 {
    let result = string1.cmp(string2);

    match result {
        std::cmp::Ordering::Less => -1,
        std::cmp::Ordering::Greater => 1,
        std::cmp::Ordering::Equal => 0,
    }
}

pub fn string_nocase_equal(string1: &str, string2: &str) -> bool {
    string_nocase_compare(string1, string2) == 0
}

pub fn string_equal<T: AsRef<str>>(string1: T, string2: T) -> bool {
    string1.as_ref() == string2.as_ref()
}

